handler/src/main/java/io/netty/handler/ssl/OptionalSslHandler.java

```diff
@@ -0,0 +1,95 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.ssl;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelHandler;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPipeline;
+import io.netty.handler.codec.ByteToMessageDecoder;
+import io.netty.util.internal.ObjectUtil;
+
+import javax.net.ssl.SSLContext;
+import java.util.List;
+
+/**
+ * {@link OptionalSslHandler} is a utility decoder to support both SSL and non-SSL handlers
+ * based on the initial message.
+ */
+public class OptionalSslHandler extends ByteToMessageDecoder {
+
+    private final SslContext sslContext;
+    private final String sslHandlerName;
+    private final String nonSslHandlerName;
+
+    public OptionalSslHandler(SslContext sslContext, String sslHandlerName, String nonSslHandlerName) {
+        this.sslContext = ObjectUtil.checkNotNull(sslContext, "sslContext");
+        this.sslHandlerName = ObjectUtil.checkNotNull(sslHandlerName, "sslHandlerName");
+        this.nonSslHandlerName = ObjectUtil.checkNotNull(nonSslHandlerName, "nonSslHandlerName");
+    }
+
+    @Override
+    protected void decode(ChannelHandlerContext context, ByteBuf in, List<Object> out) throws Exception {
+        if (in.readableBytes() < SslUtils.SSL_RECORD_HEADER_LENGTH) {
+            return;
+        }
+        if (SslHandler.isEncrypted(in)) {
+            handleSsl(context);
+        } else {
+            handleNonSsl(context);
+        }
+    }
+
+    private void handleSsl(ChannelHandlerContext context) {
+        ChannelPipeline pipeline = context.pipeline();
+        pipeline.replace(this, sslHandlerName, newSslHandler(sslContext, context));
+    }
+
+    private void handleNonSsl(ChannelHandlerContext context) {
+        ChannelHandler handler = newNonSslHandler(context);
+        if (handler != null) {
+            context.pipeline().replace(this, nonSslHandlerName, handler);
+        } else {
+            context.pipeline().remove(this);
+        }
+    }
+
+    /**
+     * Override to configure the SslHandler eg. enable HttpsEndpointIdentificationAlgorithm.
+     * @param sslContext the {@link SSLContext} to use.
+     * @param context the {@link ChannelHandlerContext} to use.
+     * @return the {@link SslHandler} to replace this handler with.
+     */
+    protected SslHandler newSslHandler(SslContext sslContext, ChannelHandlerContext context) {
+        return sslContext.newHandler(context.alloc());
+    }
+
+    /**
+     * Override to configure the ChannelHandler.
+     * @param context the {@link ChannelHandlerContext} to use.
+     * @return the {@link ChannelHandler} to replace this handler with or null to remove.
+     */
+    protected ChannelHandler newNonSslHandler(ChannelHandlerContext context) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this unused method parameter "context". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1172)